### ISSUE TITLE
Outputs of builds failing due to cyclic references not moved out of chroot

### ISSUE THREAD

**Issue by MaxHearnden** [NixOS/nix#8251]
**Describe the bug**

If a build fails due to a cycle detected in the outputs, the outputs are not moved out of the chroot

**Steps To Reproduce**

`nix build github:MaxHearnden/nix-cycle-test`
`cat $(nix eval --raw .#packages.x86_64-linux.default.out)`

**Expected behavior**

The outputs should be moved out of the chroot like when the build fails due to a non zero exit code

**`nix-env --version` output**

nix-env (Nix) 2.11.1

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by MaxHearnden** [comment]
After looking at the code, it looks like the problem might be that moving the paths out of the chroot is handled by `LocalDerivationGoal::cleanupDecideWhetherDiskFull`. This is a problem as the cyclic dependencies are detected after `LocalDerivationGoal::cleanupDecideWhetherDiskFull` might be run.

**Comment by roberth** [comment]
Would `--keep-failed` serve as a workaround?

**Comment by MaxHearnden** [comment]
At the moment the `--keep-failed` flag is not propagated to ssh-ng builders so I will wait until https://github.com/NixOS/nix/issues/5638 is resolved

**Comment by ncfavier** [comment]
Is `--keep-failed` supposed to move the outputs out of the chroot? For me it only keeps the build directory in `/tmp`. I don't know how to debug cyclic references currently.

**Comment by ncfavier** [comment]
I ended up doing something silly like

```nix
  postFixup = ''
    hash=''${dev##*/}
    hash=''${hash%%-*}
    find "$out" -type f -exec sed -i "s|$hash|GREPME|g" {} +
  '';
```

then find references from `out` to `dev` by `grep GREPME $out`.

**Comment by roberth** [comment]
> Is `--keep-failed` supposed to move the outputs out of the chroot?

I think it doesn't work for remote builds.

**Comment by ncfavier** [comment]
I was building locally!

**Comment by Ericson2314** [comment]
Yeah this issue is not about `--keep-failed` with remote building being bad: that's a separate issue. This issue is about the outputs being deleted making it hard to debug the "cycle detected..." failure.
